modules/custom/falcon_commerce/src/Plugin/rest/resource/DonationResource.php

```diff
@@ -0,0 +1,294 @@
+      watchdog_exception('falcon_commerce', $e);
+      throw new HttpException(500, 'Internal Server Error', $e);
+    }
+
+    $order_items[] = $order_item;
+
+    // Get user and profile objects.
+    $account = $this->getUser();
+    $profile = $this->getProfile();
+
+    // Create an order.
+    /** @var \Drupal\commerce_order\Entity\OrderInterface $order */
+    $order = $this->entityTypeManager->getStorage('commerce_order')->create([
+      'type' => 'donation',
+      'state' => 'draft',
+      'mail' => $account->getEmail(),
+      'uid' => $account->id(),
+      'billing_profile' => $profile,
+      'store_id' => reset($store_ids),
+      'order_items' => $order_items,
+      'placed' => $this->time->getRequestTime(),
+      'field_appeal' => $this->data['appeal'],
+    ]);
+
+    try {
+      // We need only entity level violations.
+      /** @var \Drupal\Core\Entity\EntityConstraintViolationListInterface $violations */
+      $violations = $order->validate()->getEntityViolations();
+      if ($violations->count() > 0) {
+        foreach ($violations as $violation) {
+          $this->logger->info($violation->getPropertyPath() ?: 'Entity' . ': ' . $violation->getMessage());
+        }
+
+        throw new \Exception('Order violations.');
+      }
+      $order->save();
+    }
+    catch (\Exception $e) {
+      watchdog_exception('falcon_commerce', $e);
+      throw new HttpException(500, 'Internal Server Error', $e);
+    }
+
+    $order->setOrderNumber($order->id());
+    $order->save();
+    return $order;
+  }
+
+  /**
+   * Process payment.
+   *
+   * @param \Drupal\commerce_order\Entity\OrderInterface $order
+   *   Order.
+   */
+  protected function processPayment(OrderInterface $order) {
+    // TODO: add payment processing.
+  }
+
+  /**
+   * Create a new profile from given data.
+   *
+   * @return \Drupal\profile\Entity\ProfileInterface
+   *   Profile entity.
+   */
+  private function getProfile() {
+    if (!empty($this->profile)) {
+      return $this->profile;
+    }
+
+    $user = $this->getUser();
+    $profile = $this->entityTypeManager->getStorage('profile')->create($this->data['profile'] + [
+      'type' => 'customer',
+      'uid' => $user->id(),
+    ]);
+
+    try {
+      $profile->save();
+    }
+    catch (\Exception $e) {
+      watchdog_exception('falcon_commerce', $e);
+    }
+
+    $this->profile = $profile;
+
+    return $this->profile;
+  }
+
+  /**
+   * Load existing or create a new user from given email.
+   *
+   * @return \Drupal\user\UserInterface
+   *   User object.
+   */
+  private function getUser() {
+    if (!empty($this->account)) {
+      return $this->account;
+    }
+
+    $email = $this->data['profile']['email'];
+    if ($account = user_load_by_mail($email)) {
+      $this->account = $account;
+    }
+    else {
+      $account = $this->entityTypeManager->getStorage('user')->create($this->data['profile'] + [
+        'name' => $email,
+        'mail' => $email,
+      ]);
```

Comment:
Make sure data from frontend won't set name, mail, status (other system properties?)

Reply:
Changed to array_merge().